### PR TITLE
rate limit: add computed descriptors

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Introduce computed descriptors to rate limits. Refactors the rate limit descriptor interface to accept StreamInfo and RequestHeaderMap. As a general refactor, adds an extension point for alternative rate limit descriptors to register.
Risk Level: low, not enabled unless used
Testing: unit
Docs Changes: yes
Release Notes: yes

### CODE REVIEW

File: api/envoy/config/route/v3/route_components.proto
Comment (kyessenov) [inline api/envoy/config/route/v3/route_components.proto:1728]: I decided to skip parsed expr proto for now to avoid introducing a dependency from core APIs on googleapis.

**Comment by markdroth** [comment]
I think I need some more context here.  What is this going to be used for, and how do you expect that it will be used?

**Comment by kyessenov** [comment]
@markdroth Rate limit descriptor act as categories for requests, since token buckets are applied per category. We have a requirement to use source pod labels as a category. Our extension exposes this information in a filter state, so we need a way to extract this info from the filter state (which means traverse the filter state down to a field).

The current design relies on adding more and more oneof messages here, which we can avoid by just using a generic AST from CEL.

File: api/envoy/config/route/v3/route_components.proto
Comment (gargnupur) [inline api/envoy/config/route/v3/route_components.proto:1726]: why is this a oneof?

Comment (kyessenov) [reply api/envoy/config/route/v3/route_components.proto:1726]: It's anticipating parsed_expr here which would introduce a dependency on expr proto. I can drop it since it's not needed right now.

Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto:1726]: If you think we're going to need the `oneof` later, let's keep it for now.  We can't add it later without breaking the go proto API.

Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto:1726]: Also, I believe that we are actually already using the CEL parsed expression proto in other places in the API, so this is probably not a new external dependency.  You might be able to just add it now.

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:1722]: Please add a `[#not-implemented-hide:]` comment here.

File: api/envoy/config/route/v3/route_components.proto
Comment (markdroth) [inline api/envoy/config/route/v3/route_components.proto:1768]: Please add a `[#not-implemented-hide:]` comment here.

File: api/envoy/config/route/v3/route_components.proto
Comment (kyessenov) [reply api/envoy/config/route/v3/route_components.proto:1726]: The dependency is coming from a filter. This is a core proto. Thoughts @htuch?

Comment (markdroth) [reply api/envoy/config/route/v3/route_components.proto:1726]: If we don't want to depend on the CEL protos for the xDS core protos, another option might be to make this more extensible by simply adding a `TypedExpressionConfig`.  In the long term, we could maybe even migrate the existing options here to be extensions as well.

@envoyproxy/api-shepherds, thoughts?  I wonder if we should add something to the API review checklist in #14399 that prompts people to consider using `TypedExtensionConfig` instead of a `oneof`?

Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:1726]: From an Envoy binary perspective, are we adding this to core router, or does it only apply to rate limit filter? Hard to tell from config. If core, @kyessenov would be nice to have clarity on when CEL can a reasonable number of the checkmarks in https://github.com/envoyproxy/envoy/blob/master/DEPENDENCY_POLICY.md#new-external-dependencies. Whatever is reasonable, but if this is going to become a core data plane dependency, it's probably good to use this opportunity to check in. CC @envoyproxy/dependency-shepherds 

From an API perspective, I don't feel super strongly, CEL is a reasonably standard thing. One area to have some concern around is that using CEL is similar to regex; we're introducing dynamic evaluation of some expression, which ultimately might have security and inter-request isolation concerns. I don't know if this is in any way actionable, but certainly when communicating in comments/docs its worth pointing out what the runtime implications are.

Comment (kyessenov) [reply api/envoy/config/route/v3/route_components.proto:1726]: I believe the runtime change is scoped to global and local rate limit filters. So this is purely a question of a proto dependency, to avoid parsing overhead with a parsed expression-as-proto. Good point about dynamic evaluation, I'll document that. 

Expression evaluation is linear in both input and expression size. The input here is request metadata.

Comment (mattklein123) [reply api/envoy/config/route/v3/route_components.proto:1726]: The fact that rate limit stuff is in the main router is a historical artifact of before we had per-filter route config, etc. None of this should be here and I think there is a tracking issue to move it all out, but I don't think that should block this PR.

Comment (kyessenov) [reply api/envoy/config/route/v3/route_components.proto]: OK, I think this forces moving expr package into common from extensions, since it gets pulled in the implementation.

File: api/envoy/config/route/v3/route_components.proto
Comment (kyessenov) [reply api/envoy/config/route/v3/route_components.proto:1724]: Implemented.

File: api/envoy/config/route/v3/route_components.proto
Comment (kyessenov) [reply api/envoy/config/route/v3/route_components.proto:1773]: Implemented.

**Comment by kyessenov** [comment]
The PR is ready for review.

I wonder what I should do about the layering enforcement between core and extensions. It seems that rate limit descriptors are in the core as a historic artifact, so I'd need to pull "Expr" package into `source/common` to match it because it is a dependency from the router. Is that a reasonable approach? Ideally, rate limit descriptors would be another extension point and separated from the router, but I don't see an easy way to achieve that right now.

**Comment by kyessenov** [comment]
Bringing this back to attention: I am looking how to resolve the layering violation in API with rate limit descriptors being in the core. There are a couple of ways around it:
1. Add an opaque Any extension to rate limit descriptors, register additional descriptors in rate limit extensions.
2. Keep as-is, move expr-related packages (CEL, other transitive stuff like ANTLR) as a "data-plane" dependency.

Opinions?

**Comment by mattklein123** [comment]
> Opinions?

Definitely not (2). As I already mentioned, the rate limit stuff is only in the core due to legacy reasons. It should all be removed. If you want to take one for the team we can do that (deprecate, move to filter specific route config).

If you aren't willing to do that I think (1) is our best option and we can add extensions there as @rgs1 is doing currently for access log formatters. In thinking about it we probably want an extension point here anyway, so I guess you could just go ahead and add it orthogonally to moving the stuff out of core.

**Comment by kyessenov** [comment]
Thanks! Updated the PR to address a new extension point for rate limit descriptors and added computed descriptors as an extension. I did some basic refactoring around rate limit interfaces o make it like other extension points.

File: CODEOWNERS
Comment (kyessenov) [inline CODEOWNERS:164]: @lizan Added as a code owner. Let me know if that doesn't work for you.

**Comment by kyessenov** [comment]
Windows failure appears to be identical to Wasm filter build failure (ANTLR).

File: CODEOWNERS
Comment (lizan) [reply CODEOWNERS:164]: ack, though I think `extensions/descriptors` are too generic, can we have a better name?

Comment (kyessenov) [reply CODEOWNERS:164]: We can spell it out as `rate_limit_descriptors`. Would that be better?

**Review by lizan — APPROVED** [review]

**Comment by kyessenov** [comment]
The coverage checker counts the #ifdef and NOT_REACHED_GCOVR_EXCL_LINE as no coverage for some reason. Any hint to work around it?

**Comment by kyessenov** [comment]
@lizan Could you please re-approve and merge? There's a merge conflict with https://github.com/envoyproxy/envoy/pull/14588 that I need to fix.

**Review by lizan — APPROVED** [review]

**Comment by kyessenov** [comment]
@moderation Hi, do you have a minute to approve deps change in this PR?

**Comment by moderation** [comment]
Done @kyessenov. Metadata only deps change
/lgtm deps

**Review by mattklein123 — APPROVED** [review]
Just skimmed, but this is great, thanks.

**Pull request merged**
